patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch

```diff
@@ -0,0 +1,80 @@
+         this.level.getProfiler().pop();
++        // Paper start - track when all ticks are required for the entity, e.g. after a piston push
++        if (this.isTickRequired()) {
++           this.remainingExactTicks -= 1;
++        }
++        // Paper end
+     }
+ 
+     public void setSharedFlagOnFire(boolean onFire) {
+@@ -3992,4 +3997,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
+         return ((ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
+     }
+     // Paper end
++ 
++    // Paper start - track when all ticks are required for the entity, e.g. after a piston push
++    private int remainingExactTicks = 0;
++
++    public void requestExactTicks(int ticks) {
++        this.remainingExactTicks = Math.max(ticks, this.remainingExactTicks);
++    }
++
++    public boolean isTickRequired() {
++        return (this.remainingExactTicks > 0);
++    }
++    // Peter end
+ }
+diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+index 87bedba9ab495edcce289c6665271d92b7165944..2434e82eeb8deb9c3a4820fe34840cf5c97e7610 100644
+--- a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
++++ b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+@@ -140,6 +140,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
+                                 }
+ 
+                                 entity.setDeltaMovement(e, g, h);
++                                entity.requestExactTicks(2);    // Paper - track when all ticks are required for the entity, e.g. after a piston push
+                                 break;
+                             }
+                         }
+@@ -173,6 +174,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
+     private static void moveEntityByPiston(Direction direction, Entity entity, double d, Direction direction2) {
+         NOCLIP.set(direction);
+         entity.move(MoverType.PISTON, new Vec3(d * (double)direction2.getStepX(), d * (double)direction2.getStepY(), d * (double)direction2.getStepZ()));
++        entity.requestExactTicks(10);    // Paper - track when all ticks are required for the entity, e.g. after a piston push
+         NOCLIP.set((Direction)null);
+     }
+ 
+diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
+index a08583863f9fa08016bdfc7949a273eaa4429927..24d4b5526d641088e1680ab05f52ce4af525e6e6 100644
+--- a/src/main/java/org/spigotmc/ActivationRange.java
++++ b/src/main/java/org/spigotmc/ActivationRange.java
+@@ -360,6 +360,10 @@ public class ActivationRange
+         if (entity instanceof Mob && ((Mob)entity).leashHolder instanceof Player) {
+             return true;
+         }
++        // immunize entities requesting exact ticks, e.g. being affected by a piston
```

Comment:
```suggestion
+        // Paper start - immunize entities requesting exact ticks, e.g. being affected by a piston
```

Reply:
unnecessary, this is part of an existing paper block

Reply:
Whoops, didn't see that. Is there a way for me to mark this as solved, or is that something only @petersv5 can do?

Reply:
Oops, I approved that change. Let me commit a revert to that.